### ISSUE TITLE
Localize UI to match international keyboard layouts

### ISSUE THREAD

**Issue by Hurricaaane** [UltimateHackingKeyboard/agent#710]
I am creating this issue to record my current research towards a general purpose localization of the UHK agent.

There are four issues I have identified regarding localization the UI:

- **Preference issue**: When the user does not use the US layout, the agent should be able to match the user's preferred locale
- **Preview issue:** When the user previews the keyboard layout, the keycap labels should the match the user's keyboard locale
- **Capture issue:** When the user presses a key during capture mode, the registered scancode should match the user's intended scancode
- **Categorization issue:** When the user opens the list of available scancodes, the categorization and ordering should match the user's keyboard locale

I have created a custom version of the agent for a while that is compatible with AZERTY keyboards (France), and I'd like to share my findings here.

## Preference issue
This is a question of how to adapt the UI based on the user's locale.

When to choose the user's locale:
- The UI could let the user select their own preferred layout based on a collection of known layouts (as opposed to calculating it). It might be possible to build such a layout upfront using some sort of known database.
- The UI may calculate the layout by obtaining this information from the operating system.

To this task, there is a nodejs module (MIT License) at [Microsoft/node-native-keymap](https://github.com/Microsoft/node-native-keymap).
It appears to be cross-platform.

<details><summary>When I run it on Windows with my French locale, this is the output <strong>(click to open)</strong></summary>
<pre>
{
  "Sleep": { "vkey": "VK_SLEEP", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "WakeUp": { "vkey": "VK_UNKNOWN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "KeyA": { "vkey": "VK_Q", "value": "q", "withShift": "Q", "withAltGr": "", "withShiftAltGr": ""},
  "KeyB": { "vkey": "VK_B", "value": "b", "withShift": "B", "withAltGr": "", "withShiftAltGr": ""},
  "KeyC": { "vkey": "VK_C", "value": "c", "withShift": "C", "withAltGr": "", "withShiftAltGr": ""},
  "KeyD": { "vkey": "VK_D", "value": "d", "withShift": "D", "withAltGr": "", "withShiftAltGr": ""},
  "KeyE": { "vkey": "VK_E", "value": "e", "withShift": "E", "withAltGr": "€", "withShiftAltGr": ""},
  "KeyF": { "vkey": "VK_F", "value": "f", "withShift": "F", "withAltGr": "", "withShiftAltGr": ""},
  "KeyG": { "vkey": "VK_G", "value": "g", "withShift": "G", "withAltGr": "", "withShiftAltGr": ""},
  "KeyH": { "vkey": "VK_H", "value": "h", "withShift": "H", "withAltGr": "", "withShiftAltGr": ""},
  "KeyI": { "vkey": "VK_I", "value": "i", "withShift": "I", "withAltGr": "", "withShiftAltGr": ""},
  "KeyJ": { "vkey": "VK_J", "value": "j", "withShift": "J", "withAltGr": "", "withShiftAltGr": ""},
  "KeyK": { "vkey": "VK_K", "value": "k", "withShift": "K", "withAltGr": "", "withShiftAltGr": ""},
  "KeyL": { "vkey": "VK_L", "value": "l", "withShift": "L", "withAltGr": "", "withShiftAltGr": ""},
  "KeyM": { "vkey": "VK_OEM_COMMA", "value": ",", "withShift": "?", "withAltGr": "", "withShiftAltGr": ""},
  "KeyN": { "vkey": "VK_N", "value": "n", "withShift": "N", "withAltGr": "", "withShiftAltGr": ""},
  "KeyO": { "vkey": "VK_O", "value": "o", "withShift": "O", "withAltGr": "", "withShiftAltGr": ""},
  "KeyP": { "vkey": "VK_P", "value": "p", "withShift": "P", "withAltGr": "", "withShiftAltGr": ""},
  "KeyQ": { "vkey": "VK_A", "value": "a", "withShift": "A", "withAltGr": "", "withShiftAltGr": ""},
  "KeyR": { "vkey": "VK_R", "value": "r", "withShift": "R", "withAltGr": "", "withShiftAltGr": ""},
  "KeyS": { "vkey": "VK_S", "value": "s", "withShift": "S", "withAltGr": "", "withShiftAltGr": ""},
  "KeyT": { "vkey": "VK_T", "value": "t", "withShift": "T", "withAltGr": "", "withShiftAltGr": ""},
  "KeyU": { "vkey": "VK_U", "value": "u", "withShift": "U", "withAltGr": "", "withShiftAltGr": ""},
  "KeyV": { "vkey": "VK_V", "value": "v", "withShift": "V", "withAltGr": "", "withShiftAltGr": ""},
  "KeyW": { "vkey": "VK_Z", "value": "z", "withShift": "Z", "withAltGr": "", "withShiftAltGr": ""},
  "KeyX": { "vkey": "VK_X", "value": "x", "withShift": "X", "withAltGr": "", "withShiftAltGr": ""},
  "KeyY": { "vkey": "VK_Y", "value": "y", "withShift": "Y", "withAltGr": "", "withShiftAltGr": ""},
  "KeyZ": { "vkey": "VK_W", "value": "w", "withShift": "W", "withAltGr": "", "withShiftAltGr": ""},
  "Digit1": { "vkey": "VK_1", "value": "&", "withShift": "1", "withAltGr": "", "withShiftAltGr": ""},
  "Digit2": { "vkey": "VK_2", "value": "é", "withShift": "2", "withAltGr": "~", "withShiftAltGr": ""},
  "Digit3": { "vkey": "VK_3", "value": "\"", "withShift": "3", "withAltGr": "#", "withShiftAltGr": ""},
  "Digit4": { "vkey": "VK_4", "value": "'", "withShift": "4", "withAltGr": "{", "withShiftAltGr": ""},
  "Digit5": { "vkey": "VK_5", "value": "(", "withShift": "5", "withAltGr": "[", "withShiftAltGr": ""},
  "Digit6": { "vkey": "VK_6", "value": "-", "withShift": "6", "withAltGr": "|", "withShiftAltGr": ""},
  "Digit7": { "vkey": "VK_7", "value": "è", "withShift": "7", "withAltGr": "`", "withShiftAltGr": ""},
  "Digit8": { "vkey": "VK_8", "value": "_", "withShift": "8", "withAltGr": "\\", "withShiftAltGr": ""},
  "Digit9": { "vkey": "VK_9", "value": "ç", "withShift": "9", "withAltGr": "^", "withShiftAltGr": ""},
  "Digit0": { "vkey": "VK_0", "value": "à", "withShift": "0", "withAltGr": "@", "withShiftAltGr": ""},
  "Enter": { "vkey": "VK_RETURN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Escape": { "vkey": "VK_ESCAPE", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Backspace": { "vkey": "VK_BACK", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Tab": { "vkey": "VK_TAB", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Space": { "vkey": "VK_SPACE", "value": " ", "withShift": " ", "withAltGr": "", "withShiftAltGr": ""},
  "Minus": { "vkey": "VK_OEM_4", "value": ")", "withShift": "°", "withAltGr": "]", "withShiftAltGr": ""},
  "Equal": { "vkey": "VK_OEM_PLUS", "value": "=", "withShift": "+", "withAltGr": "}", "withShiftAltGr": ""},
  "BracketLeft": { "vkey": "VK_OEM_6", "value": "^", "withShift": "¨", "withAltGr": "", "withShiftAltGr": ""},
  "BracketRight": { "vkey": "VK_OEM_1", "value": "$", "withShift": "£", "withAltGr": "¤", "withShiftAltGr": ""},
  "Backslash": { "vkey": "VK_OEM_5", "value": "*", "withShift": "µ", "withAltGr": "", "withShiftAltGr": ""},
  "Semicolon": { "vkey": "VK_M", "value": "m", "withShift": "M", "withAltGr": "", "withShiftAltGr": ""},
  "Quote": { "vkey": "VK_OEM_3", "value": "ù", "withShift": "%", "withAltGr": "", "withShiftAltGr": ""},
  "Backquote": { "vkey": "VK_OEM_7", "value": "²", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Comma": { "vkey": "VK_OEM_PERIOD", "value": ";", "withShift": ".", "withAltGr": "", "withShiftAltGr": ""},
  "Period": { "vkey": "VK_OEM_2", "value": ":", "withShift": "/", "withAltGr": "", "withShiftAltGr": ""},
  "Slash": { "vkey": "VK_OEM_8", "value": "!", "withShift": "§", "withAltGr": "", "withShiftAltGr": ""},
  "CapsLock": { "vkey": "VK_CAPITAL", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F1": { "vkey": "VK_F1", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F2": { "vkey": "VK_F2", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F3": { "vkey": "VK_F3", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F4": { "vkey": "VK_F4", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F5": { "vkey": "VK_F5", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F6": { "vkey": "VK_F6", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F7": { "vkey": "VK_F7", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F8": { "vkey": "VK_F8", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F9": { "vkey": "VK_F9", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F10": { "vkey": "VK_F10", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F11": { "vkey": "VK_F11", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F12": { "vkey": "VK_F12", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "PrintScreen": { "vkey": "VK_SNAPSHOT", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "ScrollLock": { "vkey": "VK_SCROLL", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Pause": { "vkey": "VK_NUMLOCK", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Insert": { "vkey": "VK_INSERT", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Home": { "vkey": "VK_HOME", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "PageUp": { "vkey": "VK_PRIOR", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Delete": { "vkey": "VK_DELETE", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "End": { "vkey": "VK_END", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "PageDown": { "vkey": "VK_NEXT", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "ArrowRight": { "vkey": "VK_RIGHT", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "ArrowLeft": { "vkey": "VK_LEFT", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "ArrowDown": { "vkey": "VK_DOWN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "ArrowUp": { "vkey": "VK_UP", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "NumLock": { "vkey": "VK_UNKNOWN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "NumpadDivide": { "vkey": "VK_DIVIDE", "value": "/", "withShift": "/", "withAltGr": "", "withShiftAltGr": ""},
  "NumpadMultiply": { "vkey": "VK_MULTIPLY", "value": "*", "withShift": "*", "withAltGr": "", "withShiftAltGr": ""},
  "NumpadSubtract": { "vkey": "VK_SUBTRACT", "value": "-", "withShift": "-", "withAltGr": "", "withShiftAltGr": ""},
  "NumpadAdd": { "vkey": "VK_ADD", "value": "+", "withShift": "+", "withAltGr": "", "withShiftAltGr": ""},
  "NumpadEnter": { "vkey": "VK_RETURN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Numpad1": { "vkey": "VK_END", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Numpad2": { "vkey": "VK_DOWN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Numpad3": { "vkey": "VK_NEXT", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Numpad4": { "vkey": "VK_LEFT", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Numpad5": { "vkey": "VK_CLEAR", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Numpad6": { "vkey": "VK_RIGHT", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Numpad7": { "vkey": "VK_HOME", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Numpad8": { "vkey": "VK_UP", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Numpad9": { "vkey": "VK_PRIOR", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Numpad0": { "vkey": "VK_INSERT", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "NumpadDecimal": { "vkey": "VK_DELETE", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "IntlBackslash": { "vkey": "VK_OEM_102", "value": "<", "withShift": ">", "withAltGr": "", "withShiftAltGr": ""},
  "ContextMenu": { "vkey": "VK_APPS", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Power": { "vkey": "VK_UNKNOWN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "NumpadEqual": { "vkey": "VK_CLEAR", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F13": { "vkey": "VK_F13", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F14": { "vkey": "VK_F14", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F15": { "vkey": "VK_F15", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F16": { "vkey": "VK_F16", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F17": { "vkey": "VK_F17", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F18": { "vkey": "VK_F18", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F19": { "vkey": "VK_F19", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F20": { "vkey": "VK_F20", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F21": { "vkey": "VK_F21", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F22": { "vkey": "VK_F22", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F23": { "vkey": "VK_F23", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "F24": { "vkey": "VK_F24", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Help": { "vkey": "VK_UNKNOWN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Undo": { "vkey": "VK_UNKNOWN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Cut": { "vkey": "VK_UNKNOWN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Copy": { "vkey": "VK_UNKNOWN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Paste": { "vkey": "VK_UNKNOWN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "AudioVolumeMute": { "vkey": "VK_VOLUME_MUTE", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "AudioVolumeUp": { "vkey": "VK_VOLUME_UP", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "AudioVolumeDown": { "vkey": "VK_VOLUME_DOWN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "NumpadComma": { "vkey": "VK_ABNT_C2", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "IntlRo": { "vkey": "VK_ABNT_C1", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "KanaMode": { "vkey": "VK_UNKNOWN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "IntlYen": { "vkey": "VK_UNKNOWN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Convert": { "vkey": "VK_UNKNOWN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "NonConvert": { "vkey": "VK_UNKNOWN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Lang1": { "vkey": "VK_UNKNOWN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Lang2": { "vkey": "VK_UNKNOWN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Lang3": { "vkey": "VK_UNKNOWN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Lang4": { "vkey": "VK_UNKNOWN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "ControlLeft": { "vkey": "VK_CONTROL", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "ShiftLeft": { "vkey": "VK_SHIFT", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "AltLeft": { "vkey": "VK_MENU", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "MetaLeft": { "vkey": "VK_LWIN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "ControlRight": { "vkey": "VK_CONTROL", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "ShiftRight": { "vkey": "VK_SHIFT", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "AltRight": { "vkey": "VK_MENU", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "MetaRight": { "vkey": "VK_RWIN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "MediaTrackNext": { "vkey": "VK_MEDIA_NEXT_TRACK", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "MediaTrackPrevious": { "vkey": "VK_MEDIA_PREV_TRACK", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "MediaStop": { "vkey": "VK_MEDIA_STOP", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "Eject": { "vkey": "VK_UNKNOWN", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "MediaPlayPause": { "vkey": "VK_MEDIA_PLAY_PAUSE", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "MediaSelect": { "vkey": "VK_LAUNCH_MEDIA_SELECT", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "LaunchMail": { "vkey": "VK_LAUNCH_MAIL", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "LaunchApp2": { "vkey": "VK_LAUNCH_APP2", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "LaunchApp1": { "vkey": "VK_LAUNCH_APP1", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "BrowserSearch": { "vkey": "VK_BROWSER_SEARCH", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "BrowserHome": { "vkey": "VK_BROWSER_HOME", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "BrowserBack": { "vkey": "VK_BROWSER_BACK", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "BrowserForward": { "vkey": "VK_BROWSER_FORWARD", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "BrowserStop": { "vkey": "VK_BROWSER_STOP", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "BrowserRefresh": { "vkey": "VK_BROWSER_REFRESH", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "BrowserFavorites": { "vkey": "VK_BROWSER_FAVORITES", "value": "", "withShift": "", "withAltGr": "", "withShiftAltGr": ""}
}
</pre>
</details>

Here's a few excerpts that I find interesting, because the key property has a mapping to a virtual key code, i.e. **KeyA** to **VK_Q**:
```
  "KeyA": { "vkey": "VK_Q", "value": "q", "withShift": "Q", "withAltGr": "", "withShiftAltGr": ""},
  "Digit1": { "vkey": "VK_1", "value": "&", "withShift": "1", "withAltGr": "", "withShiftAltGr": ""},
  "Digit2": { "vkey": "VK_2", "value": "é", "withShift": "2", "withAltGr": "~", "withShiftAltGr": ""},
  "Semicolon": { "vkey": "VK_M", "value": "m", "withShift": "M", "withAltGr": "", "withShiftAltGr": ""},
  "Backquote": { "vkey": "VK_OEM_7", "value": "²", "withShift": "", "withAltGr": "", "withShiftAltGr": ""},
  "IntlBackslash": { "vkey": "VK_OEM_102", "value": "<", "withShift": ">", "withAltGr": "", "withShiftAltGr": ""},
```

It might be possible to use this module either on the user's machine, or in order to pre-build a collection of known layouts.

There is also [this documentation from archlinux wiki](https://wiki.archlinux.org/index.php/Map_scancodes_to_keycodes) but I haven't explored it much.

## Preview issue
Regarding the preview issue, my current solution is to use the English layout as a base for the French locale.
Then I'm overriding part of the existing labels with known differences from the French locale, like a stencil.

![example1](https://user-images.githubusercontent.com/839160/42134327-e4718362-7d39-11e8-8eeb-0fcb75f51cca.png)

When I made this, I did not use node-native-keymap, it was done by hand.

**If the node-native-keymap module is consistent enough**, it might be possible to automate figuring out keycap labels and/or calculate it at runtime.

Example commit: https://github.com/Hurricaaane/agent/commit/e463e897999a6b3dac8d27e7b1c13989b65678e8#diff-b587572d474fcffcd44df49caaf6695eR13

## Capture issue

During my research, I found out that it's possible from JavaScript to [record the physical position of the key](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code)
rather than a keyCode, which could tremendously help with the Capture functionnality, but according to the same source, it doesn't work on Mac OS X.

For my own solution, I'm using the English keyCode to scancode conversion as a base for the French capture.
Then I'm overriding part of the mapping with my own locale.

In the following screenshot, it means that when I press the key that I have already configured, the captured key does not change.
On top is my implementation, on bottom is the official agent.

![example2](https://user-images.githubusercontent.com/839160/42134326-e454f332-7d39-11e8-8da4-fba40aec74da.png)

**If the node-native-keymap module is consistent enough**, it might be possible to use a reverse keymap to convert the recorded keyCode into the proper scancode. For instance:
- The user who uses a French keyboard layout on their OS, presses the key Q on their keyboard during capture mode
- Javascript has recorded an event with keyCode 81
- keyCode 81 is mapped to the Virtual Key VK_Q
- The user's locale lookup table maps Virtual Key VK_Q to KeyA
```
  "KeyA": { "vkey": "VK_Q", "value": "q", "withShift": "Q", "withAltGr": "", "withShiftAltGr": ""},
```
- KeyA is mapped to scancode 4, which is then stored in the configuration

Example commit: https://github.com/Hurricaaane/agent/commit/dfdcfa81894a5c82702e7147d5a1514a1110408c#diff-b587572d474fcffcd44df49caaf6695eR35

## Categorization issue

The menu is locale-dependant.

For instance, in English, the [scancode 16 is the letter M](https://github.com/Hurricaaane/agent/blob/d3b08111ca42c467ad99afbddf5f142a124bb000/packages/uhk-web/src/app/locales/scancodes/scancodes.common.ts#L68
).

In French, the [scancode 16 are question mark and comma](https://github.com/Hurricaaane/agent/blob/d3b08111ca42c467ad99afbddf5f142a124bb000/packages/uhk-web/src/app/locales/scancodes/scancodes.fr.ts#L17).

This causes two issues:
- Categorization, as some scancodes are not in the same category depending on the locale, and may belong to multiple categories
- Ordering of the scancodes in the menu needs to be adjusted as some letters are swapped

I have addressed this by applying tags to keys, but also choosing a different ordering scheme based on their category.

Example commit: https://github.com/Hurricaaane/agent/commit/dfdcfa81894a5c82702e7147d5a1514a1110408c#diff-b587572d474fcffcd44df49caaf6695eR35

**Comment by Hurricaaane** [comment]
I'll try to work towards a general purpose localization solution, but I can't make any guarantees.

This is a POC that I had kept for myself since I wasn't able to allocate free time to study a general purpose solution. I think it may be beneficial for some UHK users if I could come up with a working solution.

**Comment by mondalaci** [comment]
I appreciate that you've taken time to do the above research and describe your findings, but according to the "Accented characters in Agent" section of my [How can I type accented characters with my UHK?](https://ultimatehackingkeyboard.com/blog/2018/06/23/how-can-i-type-accented-characters-with-my-uhk) blost post, I'm fully confident that the implementation effort and long-term mainteance burden of this feature would be vast compared to the value it provides.

Your research includes some concrete solutions but I don't think any of them would work completely and not in a cross-platform manner, and even if they would, the fundamental difference between scancodes and characters would confuse users when moving their keyboards across computers of different OS layouts.

The bottom line is that we're a small team, and there are still a ton of issues we have to deal with. I don't want to add a feature that is so incredibly problematic and which adds a huge long-term mainteance and support burden.

Feel free to follow up in this issue, but for the time being, I'm closing it.
